ddl/pause_test.go

```diff
@@ -0,0 +1,306 @@
+	// Modify column, no reorg.
+	{"alter table t_user modify column c11 mediumint", true, model.StateNone, true, false, nil},
+	{"alter table t_user modify column c11 int", false, model.StatePublic, false, true, nil},
+
+	// Modify column, reorg.
+	{"alter table t_user modify column c11 char(10)", true, model.StateNone, true, false, nil},
+	{"alter table t_user modify column c11 char(10)", true, model.StateDeleteOnly, true, true, nil},
+	{"alter table t_user modify column c11 char(10)", true, model.StateWriteOnly, true, true, nil},
+	{"alter table t_user modify column c11 char(10)", true, model.StateWriteReorganization, true, true, nil},
+	{"alter table t_user modify column c11 char(10)", false, model.StatePublic, false, true, nil},
+}
+
+func isCommandSuccess(rs *testkit.Result) bool {
+	return strings.Contains(rs.Rows()[0][1].(string), "success")
+}
+
+func TestPauseAndResumeMain(t *testing.T) {
+	store, dom := testkit.CreateMockStoreAndDomainWithSchemaLease(t, 100*time.Millisecond)
+	tk := testkit.NewTestKit(t, store)
+	tkCommand := testkit.NewTestKit(t, store)
+
+	tk.MustExec("use test")
+	tk.MustExec(`CREATE TABLE if not exists t_user (
+        id int(11) NOT NULL AUTO_INCREMENT,
+        user varchar(128) NOT NULL,
+        name varchar(128) NOT NULL,
+        age int(11) NOT NULL,
+        province varchar(32) NOT NULL DEFAULT '',
+        city varchar(32) NOT NULL DEFAULT '',
+        phone varchar(16) NOT NULL DEFAULT '',
+        created_time datetime NOT NULL,
+        updated_time datetime NOT NULL,
+        PRIMARY KEY (id)
+      ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;`)
+
+	idx := 0
+	rowCount := 100000
+	tu := &TestTableUser{}
+	for idx < rowCount {
+		_ = tu.generateAttributes()
+		tk.MustExec(tu.insertStmt())
+
+		idx += 1
+	}
+
+	ddl.ReorgWaitTimeout = 10 * time.Millisecond
+	tk.MustExec("set @@global.tidb_ddl_reorg_batch_size = 2")
+	tk.MustExec("set @@global.tidb_ddl_reorg_worker_cnt = 1")
+	tk = testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/ddl/mockBackfillSlow", "return"))
+	defer func() {
+		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/ddl/mockBackfillSlow"))
+	}()
+
+	hook := &callback.TestDDLCallback{Do: dom}
+	i := atomicutil.NewInt64(0)
+
+	isPaused := atomicutil.NewBool(false)
+	pauseWhenReorgNotStart := atomicutil.NewBool(false)
+	pauseHook := func(job *model.Job) {
+		if testMatchCancelState(t, job, allPauseJobTestCase[i.Load()].jobState, allPauseJobTestCase[i.Load()].sql) && !isPaused.Load() {
+			if !pauseWhenReorgNotStart.Load() && job.SchemaState == model.StateWriteReorganization && job.MayNeedReorg() && job.RowCount == 0 {
+				return
+			}
+			rs := tkCommand.MustQuery(fmt.Sprintf("admin pause ddl jobs %d", job.ID))
+			isPaused.Store(isCommandSuccess(rs))
+		}
+	}
+
+	isResumed := atomicutil.NewBool(false)
+	resumeHook := func(job *model.Job) {
+		rs := tkCommand.MustQuery(fmt.Sprintf("admin cancel ddl jobs %d", job.ID))
```

Comment:
Do we need to change it to `resume`?

Reply:
Should not test it here and now. Add it later together with the reorganization change.